kpi/deployment_backends/kobocat_backend.py

```diff
@@ -491,47 +493,57 @@ def generate_new_instance_id() -> (str, str):
         return _uuid, f'uuid:{_uuid}'
 
     def get_attachment_content(
-        self, submission_id: int, user: 'auth.User', xpath: str
+        self,
+        submission_id: int,
+        user: 'auth.User',
+        attachment_id: Optional[int] = None,
+        xpath: Optional[str] = None,
     ) -> tuple:
+        """
+        Return a tuple which contains the filename of the attachment, its content
+        and its mimetype.
+        Attachment can be retrieved by its primary key or by XPath.
+        An exception is raised when the submission or the attachment is not found.
+        """
+
         submission_xml = self.get_submission(
             submission_id, user, format_type=SUBMISSION_FORMAT_TYPE_XML
         )
 
         if not submission_xml:
-            raise Http404
+            raise SubmissionNotFoundException
+
+        if xpath:
+            submission_tree = ET.ElementTree(ET.fromstring(submission_xml))
+            element = submission_tree.find(xpath)
 
-        submission_tree = ET.ElementTree(ET.fromstring(submission_xml))
-        element = submission_tree.find(xpath)
+            try:
+                attachment_filename = element.text
+            except AttributeError:
+                raise InvalidXPathException
+            filters = {
+                'instance_id': submission_id,
+                'media_file_basename': attachment_filename,
+            }
+        else:
+            filters = {
+                'instance_id': submission_id,
+                'pk': attachment_id,
+            }
 
         try:
-            attachment_filename = element.text
-        except AttributeError:
-            raise InvalidXPathException
-
-        # No need to validate is `submission_json` equals `None`. It has
-        # already be done with `submission_xml`.
-        submission_json = self.get_submission(
-            submission_id, user, format_type=SUBMISSION_FORMAT_TYPE_JSON
-        )
+            attachment = ReadOnlyKobocatAttachment.objects.get(**filters)
```

Comment:
Oli says: this won't work for the first attempt at accessing small/medium/large resized images

Reply:
#3672